docs/book/provider_implementations/populate_controllers.md

```diff
@@ -0,0 +1,159 @@
+# Controllers and Reconciliation
+
+From the [kubebuilder book][controller]:
+
+> Controllers are the core of Kubernetes, and of any operator.
+>
+> It’s a controller’s job to ensure that, for any given object, the actual state of the world (both the cluster state, and potentially external state like running containers for Kubelet or loadbalancers for a cloud provider) matches the desired state in the object. 
+> Each controller focuses on one root Kind, but may interact with other Kinds.
+>
+> We call this process reconciling.
+
+[controller]: https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller
+
+Right now, we can create objects in our API but we won't do anything about it. Let's fix that. 
+
+# Let's see the Code
+
+Kubebuilder has created our first controller in `controllers/mailguncluster_controller.go`. Let's take a look at what got generated:
+
+```go
+// MailgunClusterReconciler reconciles a MailgunCluster object
+type MailgunClusterReconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=mailgunclusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=mailgunclusters/status,verbs=get;update;patch
+
+func (r *MailgunClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("mailguncluster", req.NamespacedName)
+
+	// your logic here
+
+	return ctrl.Result{}, nil
+}
+```
+
+## State
+
+Let's focus on that `struct` first. 
+First, a word of warning: no guarantees are made about parallel access, both on one machine or multiple machines.
+That means you should not store any important state in memory: if you need it, write it into a Kubernetes object and store it.
+
+We're going to be sending mail, so let's add a few extra fields:
+
+// MailgunClusterReconciler reconciles a MailgunCluster object
+type MailgunClusterReconciler struct {
+	client.Client
+	Log       logr.Logger
+	Mailgun   mailgun.Mailgun
+	Recipient string
+}
+
+## Reconciliation
+
+Now it's time for our Reconcile function.
+Reconcile is only passed a name, not an object, so let's retrieve ours.
+
+Here's a naive example:
+
+```
+func (r *MailgunClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	_ = r.Log.WithValues("mailguncluster", req.NamespacedName)
```

Comment:
You'll need to store this as a variable (`log` or `logger`) to be able to log - this line is a no-op as written.